src/charm.py

```diff
@@ -6,38 +6,197 @@
+            )
+        return self._lightkube_client
+
+    def _get_waypoint_resource_manager(self):
+        return KubernetesResourceManager(
+            labels=create_charm_default_labels(
+                self.app.name, self.model.name, scope=WAYPOINT_LABEL
+            ),
+            resource_types=WAYPOINT_RESOURCE_TYPES,  # pyright: ignore
+            lightkube_client=self.lightkube_client,
+            logger=logger,
+        )
+
+    def _is_deployment_ready(self) -> bool:
+        """Check if the deployment is ready after 10 attempts."""
+        timeout = int(self.config["ready-timeout"])
+        check_interval = 10
+        attempts = timeout // check_interval
+
+        for _ in range(attempts):
+            try:
+                deployment = self.lightkube_client.get(
+                    Deployment,
+                    name=f"{self.app.name}-{self.model.name}-waypoint",
+                    namespace=self.model.name,
+                )
+                if (
+                    deployment.status
+                    and deployment.status.readyReplicas == deployment.status.replicas
+                ):
+                    return True
+            except ApiError as e:
+                logger.error(f"Error checking waypoint deployment status: {e}")
 
-    def update_mesh(self):
-        """Update the service mesh.
+            time.sleep(check_interval)
 
-        Reads all relations and any config to generate a kubernetes manifest. Then applies the
-        manifest.
+        return False
+
+    def _is_ready(self) -> bool:
+
+        if not self._is_deployment_ready():
+            return False
+        return True
+
+    def _sync_all_resources(self):
+        self.unit.status = MaintenanceStatus("Validating waypoint readiness")
+        self._sync_waypoint_resources()
+        if not self._is_ready():
+            self.unit.status = BlockedStatus(
+                "Waypoint's k8s deployment not ready, is istio properly installed?"
+            )
+            return
+        self.unit.status = ActiveStatus()
+
+    def _construct_waypoint(self):
+        gateway = IstioWaypointResource(
+            metadata=Metadata(
+                name=f"{self.app.name}-{self.model.name}-waypoint",
+                namespace=self.model.name,
+                labels={"istio.io/waypoint-for": "service"},
+            ),
+            spec=IstioWaypointSpec(
+                gatewayClassName="istio-waypoint",
+                listeners=[
+                    Listener(
+                        name="mesh",
+                        port=15008,
+                        protocol="HBONE",
+                        allowedRoutes=AllowedRoutes(namespaces={"from": "All"}),
+                    )
+                ],
+            ),
+        )
+        gateway_resource = RESOURCE_TYPES["Gateway"]
+        return gateway_resource(
+            metadata=ObjectMeta.from_dict(gateway.metadata.dict()),
+            spec=gateway.spec.dict(),
```

Comment:
dict() is deprecated
```suggestion
            spec=gateway.spec.model_dump(),
```